### PR TITLE
Immutable dataset

### PR DESCRIPTION
**Pull request by ramicetty**
--> As of now we can take input as array from user.In future(next step) List type.
--> If given array having single value then creating SingletonList.
--> if array is empty then returning NullList( in next step).
--> As of now tuple(Finite List) size is 32.we can increase the size as per our requirement.
--> In future(next step) we can create infinite list as well as.
--> As of now i created include method to adding one more element into list.
Kindly go through the code and give your comments plzz.

### CODE REVIEW

File: src/main/java/com/jprobstats/dataset/impl/Tuple.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+		for (int i = 0; i < n; i++) {
+			ndata[i] = values[i + fromIndex];
+		}
+		return new Tuple<T>(ndata);
+	}
+
+	/**
+	 * @return size of the data set.
+	 */
+	public int size() {
```
Comment (manojiitkumar93) [inline src/main/java/com/jprobstats/dataset/impl/Tuple.java]: Constructor is made private, then i think this method should be static...
Correct me if I am wrong.

File: src/main/java/com/jprobstats/dataset/impl/Tuple.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+		return data.length;
+	}
+
+	/**
+	 * @return value from data set based on given index.
+	 */
+	public T get(int i) {
```
Comment (manojiitkumar93) [inline src/main/java/com/jprobstats/dataset/impl/Tuple.java]: Constructor is made private, then i think this method should be static...
Correct me if I am wrong.

Comment (manojiitkumar93) [inline src/main/java/com/jprobstats/dataset/impl/Tuple.java]: check for the index here...
I mean wt if index is greater than size of the data

File: src/main/java/com/jprobstats/dataset/ListFactory.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	 * @param toIndex
+	 * @return
+	 */
+	public static <T> JProbStatsList<T> createFromArray(T[] data, int fromIndex, int toIndex) {
+		int n = toIndex - fromIndex;
+		// TODO if given array size is more than tuple(finite size) then
+		// create BlockList with infinite size
+		if (n <= MAX_TUPLE_BUILD_SIZE) {
+			// TODO if n==0 then return empty List(NullList).
+			if (n < 2) {
```
Comment (manojiitkumar93) [inline src/main/java/com/jprobstats/dataset/ListFactory.java]: I think even without this if loop, we can achieve our logic...

File: src/main/java/com/jprobstats/dataset/impl/Tuple.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+		return data[i];
+	}
+
+	@SuppressWarnings("unchecked")
+	public static <T> Tuple<T> concat(JProbStatsList<T> a, JProbStatsList<T> b) {
```
Comment (manojiitkumar93) [inline src/main/java/com/jprobstats/dataset/impl/Tuple.java]: is it fine if both inputs "a" and "b" are empty/null.....

Comment (manojiitkumar93) [inline src/main/java/com/jprobstats/dataset/impl/Tuple.java]: If I want to concat more than 2 Lists do i need to call this "concat" method multiple times?

File: src/main/java/com/jprobstats/dataset/impl/CompositeList.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+	   return new CompositeList<T>(a,b);
+	}
+	
+	@Override
+	public T get(int i) {
+		int fs = front.size();
+		if (i < fs) {
+			return front.get(i);
+		}
+		return back.get(i - fs);
```
Comment (manojiitkumar93) [inline src/main/java/com/jprobstats/dataset/impl/CompositeList.java]: if front size is 2 and back size is 3, then the size of compositeList will be 5. Now wt if we call get(int i) will i=6, it throws exception, can you handle this case....

File: src/main/java/com/jprobstats/dataset/impl/CompositeList.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+	public final JProbStatsList<T> back;
+	private final int size;
+
+	private CompositeList(JProbStatsList<T> a, JProbStatsList<T> b) {
+		front = a;
+		back = b;
+		size = a.size() + b.size();
+	}
+	// TODO when both list size exceeds tuple size we have to increase destination array.
+	public static <T> JProbStatsList<T> concat(JProbStatsList<T> a, JProbStatsList<T> b) {
```
Comment (manojiitkumar93) [inline src/main/java/com/jprobstats/dataset/impl/CompositeList.java]: If any one of the input is null, then a.size() or b.size() will throw null pointer exception. You may want to look into this once again...

File: src/main/java/com/jprobstats/dataset/impl/CompositeList.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+		int as=a.size(); if (as==0) return b;
+		int bs=b.size(); if (bs==0) return a;
+		if ((as+bs)<=ListFactory.MAX_TUPLE_BUILD_SIZE) {
+			return Tuple.concat(a, b);
```
Comment (manojiitkumar93) [inline src/main/java/com/jprobstats/dataset/impl/CompositeList.java]: Similarly we have concat method in Tuple class, there it just combines to one list and returns new tuple. 
So do i need CompositeList class? 
Can u provide some uses cases of it and what will be the difference Tuple.concat and CompositeList.concat ....

Ignore if my questions are invalid....

File: src/main/java/com/jprobstats/dataset/impl/CompositeList.java (hunk 6 above)
Comment (ramicetty) [inline src/main/java/com/jprobstats/dataset/impl/CompositeList.java]: If my dataset is null list then size will be 0.it wont throw null ponter exception.

File: src/main/java/com/jprobstats/dataset/impl/CompositeList.java (hunk 5 above)
Comment (ramicetty) [inline src/main/java/com/jprobstats/dataset/impl/CompositeList.java]: Yes you are correct. if given index is less than 0 and grater than data set  size i have to throw exception.Thanks for making me correct.

File: src/main/java/com/jprobstats/dataset/impl/Tuple.java (hunk 4 above)
Comment (ramicetty) [inline src/main/java/com/jprobstats/dataset/impl/Tuple.java]: As of now i am giving option to user like they can  append one value,one array or one list with existing data set.

In this concat method,one list is already available data other list is which user wants to append.

Comment (ramicetty) [inline src/main/java/com/jprobstats/dataset/impl/Tuple.java]: if both lists are empty/null then no issue. my destination data set will also null.

Example :: ListFactory.createFromArray(new Integer[] {}); // my data set will be null list with size 0.
 ListFactory.createFromArray(new Integer[] {null}); // my data set will contain null value with size 1.
Note :: My data set must accept null values as well.
Example :: students marks:: 80 70 60 null 90

Here null means one student did't write the exam.

File: src/main/java/com/jprobstats/dataset/ListFactory.java (hunk 3 above)
Comment (ramicetty) [inline src/main/java/com/jprobstats/dataset/ListFactory.java]: may i know which point of view  you are telling

File: src/main/java/com/jprobstats/dataset/impl/Tuple.java (hunk 2 above)
Comment (ramicetty) [inline src/main/java/com/jprobstats/dataset/impl/Tuple.java]: Yes,We have to check.thanks for making me correct.

File: src/main/java/com/jprobstats/dataset/impl/CompositeList.java (hunk 7 above)
Comment (ramicetty) [inline src/main/java/com/jprobstats/dataset/impl/CompositeList.java]: yes you are correct. for readability i divided like this.

File: src/main/java/com/jprobstats/dataset/DataFactory.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
+package com.jprobstats.dataset;
+
+import java.util.List;
+
+import com.jprobstats.dataset.impl.CompositeData;
+import com.jprobstats.dataset.impl.DataSet;
+import com.jprobstats.dataset.impl.NullData;
+import com.jprobstats.dataset.impl.SingletonData;
+
+public class DataFactory<T> {
```
Comment (UTTAM2311) [inline src/main/java/com/jprobstats/dataset/DataFactory.java:10]: may the generic be a bounded one like T extends Number

File: src/main/java/com/jprobstats/dataset/IData.java [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+
+/**
+ * we can add some more methods in future
+ *
+ * @param <T>
+ */
+public interface IData<T> {
+
+	// access methods
+
```
Comment (mayank12001) [inline src/main/java/com/jprobstats/dataset/IData.java:11]: aren't we restricting data to be Integer only ?

File: src/main/java/com/jprobstats/dataset/impl/SubList.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+		offset = off;
+		length = len;
+	}
+
+	@SuppressWarnings({ "unchecked", "rawtypes" })
+	public static final SubList<?> EMPTY_SUBLIST = new SubList(DataFactory.emptyList(), 0, 0);
+
+	@SuppressWarnings("unchecked")
+	public static <T> SubList<T> create(List<T> source, int fromIndex, int toIndex) {
+		if ((fromIndex < 0) || (toIndex > source.size()))
```
Comment (ghost) [inline src/main/java/com/jprobstats/dataset/impl/SubList.java:25]: condition should be `toIndex >= source.size()`

**Pull request merged**